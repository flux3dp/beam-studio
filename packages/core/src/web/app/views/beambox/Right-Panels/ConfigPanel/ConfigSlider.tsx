import classNames from 'classnames';
import React, { memo, useCallback, useEffect, useMemo, useState } from 'react';
import { ConfigProvider, Slider, TooltipProps } from 'antd';

import ConfigOption from 'interfaces/ConfigOption';
import units from 'helpers/units';

import styles from './ConfigSlider.module.scss';

interface Props {
  id?: string;
  value: number;
  onChange: (value: number) => void;
  min: number;
  max: number;
  step?: number;
  speedLimit?: boolean;
  options?: ConfigOption[];
  unit?: string;
  decimal?: number;
}

const ConfigSlider = ({
  id,
  value,
  onChange,
  min,
  max,
  step = 1,
  speedLimit = false,
  options,
  unit,
  decimal = 0,
}: Props) => {
  // If value is not in options, add the value to options
  const sliderOptions = useMemo(() => {
    if (!options) return undefined;
    const newOptions = [...options];
    for (let i = 0; i < options.length; i += 1) {
      if (options[i].value === value) return newOptions;
      if (options[i].value > value) {
        newOptions.splice(i, 0, { value, label: `${value}` });
        return newOptions;
      }
    }
    newOptions.push({ value, label: `${value}` });
    return newOptions;
  }, [value, options]);

  const optionValues = useMemo(() => sliderOptions?.map((option) => option.value), [sliderOptions]);
  const optionLabels = useMemo(
    () => sliderOptions?.map((option) => option.label ?? option.value),
    [sliderOptions]
  );
  const fakeUnit = useMemo(() => (unit?.includes('in') ? 'inch' : 'mm'), [unit]);
  const getDisplayValueFromValue = useCallback(
    (val: number) => {
      if (optionValues) return optionValues.indexOf(val);
      return val;
    },
    [optionValues]
  );

  const [displayValue, setDisplayValue] = useState(getDisplayValueFromValue(value));
  useEffect(
    () => setDisplayValue(getDisplayValueFromValue(value)),
    [value, getDisplayValueFromValue]
  );

  const handleAfterChange = (val: number) => {
    if (optionValues) onChange(optionValues[val]);
    else onChange(val);
  };

  const handleChange = (val: number) => {
    setDisplayValue(val);
  };
  const maxValue = sliderOptions ? sliderOptions.length - 1 : max;
  return (
    <div id={id} className={classNames(styles.container, { [styles.limit]: speedLimit })}>
      <ConfigProvider
        theme={{
          token: {
            colorPrimaryBorder: '#cecece',
            colorPrimaryBorderHover: '#494949',
            colorPrimary: '#494949',
          },
          components: {
            Slider: {
              handleColor: '#cecece',
              handleActiveColor: '#494949',
              dotActiveBorderColor: '#494949',
              trackBg: 'transparent',
              trackBgDisabled: 'transparent',
              trackHoverBg: 'transparent',
            },
          },
        }}
      >
        <Slider
          min={sliderOptions ? 0 : min}
          max={sliderOptions ? sliderOptions.length - 1 : max}
          step={sliderOptions ? 1 : step}
          value={displayValue}
          onAfterChange={handleAfterChange}
          onChange={handleChange}
          tooltip={
            {
              formatter: (val: number) =>
                sliderOptions
                  ? optionLabels[val]
                  : units.convertUnit(val, fakeUnit, 'mm').toFixed(decimal),
              // hack because antd tooltip of slider won't autoslide
              placement: displayValue === maxValue ? 'topLeft' : 'top',
              arrow: { pointAtCenter: true },
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
            } as TooltipProps as any
          }
        />
      </ConfigProvider>
    </div>
  );
};

export default memo(ConfigSlider);
