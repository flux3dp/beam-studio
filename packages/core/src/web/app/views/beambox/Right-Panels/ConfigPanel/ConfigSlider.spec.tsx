/* eslint-disable @typescript-eslint/no-explicit-any */ // for mock components props
import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import ConfigSlider from './ConfigSlider';

const MOCK_CHANGE_VALUE = 88;
const MOCK_AFTER_CHANGE_VALUE = 87;

jest.mock('antd', () => ({
  ConfigProvider: ({ theme, children }: any) => (
    <div>
      Mock Antd ConfigProvider<p>theme: {JSON.stringify(theme)}</p>
      {children}
    </div>
  ),
  Slider: ({ min, max, step, value, onAfterChange, onChange }: any) => (
    <div>
      Mock Antd Slider
      <p>min: {min}</p>
      <p>max: {max}</p>
      <p>step: {step}</p>
      <p>value: {value}</p>
      <button type="button" onClick={() => onAfterChange(MOCK_AFTER_CHANGE_VALUE)}>
        onAfterChange
      </button>
      <button type="button" onClick={() => onChange(MOCK_CHANGE_VALUE)}>
        onChange
      </button>
    </div>
  ),
}));

const mockPropOnChange = jest.fn();

describe('test ConfigSlider', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should work', () => {
    const { container, getByText, rerender } = render(
      <ConfigSlider
        id="mock-id"
        value={10}
        onChange={mockPropOnChange}
        min={0}
        max={100}
        step={1}
      />
    );
    expect(container).toMatchSnapshot();
    rerender(
      <ConfigSlider
        id="mock-id"
        value={7}
        onChange={mockPropOnChange}
        min={0}
        max={100}
        step={1}
        speedLimit
      />
    );
    expect(container).toMatchSnapshot();
    expect(getByText('value: 7')).toBeInTheDocument();
    fireEvent.click(getByText('onChange'));
    expect(container).toMatchSnapshot();
    expect(getByText(`value: ${MOCK_CHANGE_VALUE}`)).toBeInTheDocument();
    expect(mockPropOnChange).not.toBeCalled();
    fireEvent.click(getByText('onAfterChange'));
    expect(mockPropOnChange).toBeCalledTimes(1);
    expect(mockPropOnChange).toHaveBeenLastCalledWith(MOCK_AFTER_CHANGE_VALUE);
  });
});
