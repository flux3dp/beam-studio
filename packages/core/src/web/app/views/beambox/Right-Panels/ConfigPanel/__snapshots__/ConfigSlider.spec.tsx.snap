// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ConfigSlider should work 1`] = `
<div>
  <div
    class="container"
    id="mock-id"
  >
    <div>
      Mock Antd ConfigProvider
      <p>
        theme: 
        {"token":{"colorPrimaryBorder":"#cecece","colorPrimaryBorderHover":"#494949","colorPrimary":"#494949"},"components":{"Slider":{"handleColor":"#cecece","handleActiveColor":"#494949","dotActiveBorderColor":"#494949","trackBg":"transparent","trackBgDisabled":"transparent","trackHoverBg":"transparent"}}}
      </p>
      <div>
        Mock Antd Slider
        <p>
          min: 
          0
        </p>
        <p>
          max: 
          100
        </p>
        <p>
          step: 
          1
        </p>
        <p>
          value: 
          10
        </p>
        <button
          type="button"
        >
          onAfterChange
        </button>
        <button
          type="button"
        >
          onChange
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`test ConfigSlider should work 2`] = `
<div>
  <div
    class="container limit"
    id="mock-id"
  >
    <div>
      Mock Antd ConfigProvider
      <p>
        theme: 
        {"token":{"colorPrimaryBorder":"#cecece","colorPrimaryBorderHover":"#494949","colorPrimary":"#494949"},"components":{"Slider":{"handleColor":"#cecece","handleActiveColor":"#494949","dotActiveBorderColor":"#494949","trackBg":"transparent","trackBgDisabled":"transparent","trackHoverBg":"transparent"}}}
      </p>
      <div>
        Mock Antd Slider
        <p>
          min: 
          0
        </p>
        <p>
          max: 
          100
        </p>
        <p>
          step: 
          1
        </p>
        <p>
          value: 
          7
        </p>
        <button
          type="button"
        >
          onAfterChange
        </button>
        <button
          type="button"
        >
          onChange
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`test ConfigSlider should work 3`] = `
<div>
  <div
    class="container limit"
    id="mock-id"
  >
    <div>
      Mock Antd ConfigProvider
      <p>
        theme: 
        {"token":{"colorPrimaryBorder":"#cecece","colorPrimaryBorderHover":"#494949","colorPrimary":"#494949"},"components":{"Slider":{"handleColor":"#cecece","handleActiveColor":"#494949","dotActiveBorderColor":"#494949","trackBg":"transparent","trackBgDisabled":"transparent","trackHoverBg":"transparent"}}}
      </p>
      <div>
        Mock Antd Slider
        <p>
          min: 
          0
        </p>
        <p>
          max: 
          100
        </p>
        <p>
          step: 
          1
        </p>
        <p>
          value: 
          88
        </p>
        <button
          type="button"
        >
          onAfterChange
        </button>
        <button
          type="button"
        >
          onChange
        </button>
      </div>
    </div>
  </div>
</div>
`;
